nucypher/blockchain/eth/actors.py

```diff
@@ -585,6 +596,118 @@ def encrypt_threshold_decryption_response(
             requester_public_key=requester_public_key,
         )
 
+
+    def _verify_active_ritual(self, decryption_request: ThresholdDecryptionRequest):
+        # TODO: #3052 consider using the DKGStorage cache instead of the coordinator agent
+        # dkg_public_key = this_node.dkg_storage.get_public_key(decryption_request.ritual_id)
+        ritual = self.coordinator_agent.get_ritual(
+            decryption_request.ritual_id, with_participants=True
+        )
+
+        # enforces that the node is part of the ritual
+        participants = [p.provider for p in ritual.participants]
+        if self.checksum_address not in participants:
+            raise self.UnauthorizedRequest(
+                f"Node not part of ritual {decryption_request.ritual_id}",
+            )
+
+        # check that ritual not timed out
+        current_timestamp = self.coordinator_agent.blockchain.get_blocktime()
+        if current_timestamp > ritual.end_timestamp:
+            raise self.UnauthorizedRequest(
+                f"Ritual {decryption_request.ritual_id} is expired",
+            )
+
+    def _verify_ciphertext_authorization(
+        self, decryption_request: ThresholdDecryptionRequest
+    ) -> None:
+        """check that the ciphertext is authorized for this ritual"""
+        ciphertext_header = decryption_request.ciphertext_header
+        authorization = decryption_request.acp.authorization
+        if not self.coordinator_agent.is_encryption_authorized(
+            ritual_id=decryption_request.ritual_id,
+            evidence=authorization,
+            ciphertext_header=bytes(ciphertext_header),
+        ):
+            raise self.UnauthorizedRequest(
+                f"Encrypted data not authorized for ritual {decryption_request.ritual_id}",
+            )
+
+    def _evaluate_conditions(
+        self, decryption_request: ThresholdDecryptionRequest
+    ) -> None:
+        # requester-supplied condition eval context
+        context = None
+        if decryption_request.context:
+            # nucypher_core.Context -> str -> dict
+            context = json.loads(str(decryption_request.context)) or dict()
+
+        # obtain condition from request
+        condition_lingo = json.loads(
+            str(decryption_request.acp.conditions)
+        )  # nucypher_core.Conditions -> str -> Lingo
+        if not condition_lingo:
+            # this should never happen for CBD - defeats the purpose
+            raise self.UnauthorizedRequest(
+                "No conditions present for ciphertext - invalid for CBD functionality",
+            )
+
+        # evaluate the conditions for this ciphertext
+        error = evaluate_condition_lingo(
+            condition_lingo=condition_lingo,
+            context=context,
+            providers=self.condition_providers,
+        )
+        if error:
```

Comment:
We need to be careful here. There is some rich information in this `error`. It is an `EvalError` type that includes a recommended status code. Many are BAD_REQUEST, but some are FORBIDDEN, NOT_IMPLEMENTED eg. https://github.com/nucypher/nucypher/blob/v7.0.0/nucypher/policy/conditions/utils.py#L85.

Perhaps we need to also update what `evaluate_condition_lingo` does/returns in the case of a problem...

Reply:
Pushed an update that allows `EvalError` to bubble up the the server routing.